src/events/attributes.rs

```diff
@@ -796,6 +892,94 @@ mod xml {
     use super::*;
     use pretty_assertions::assert_eq;
 
+    #[cfg(any(doc, not(feature = "encoding")))]
+    mod attribute_value_normalization {
+
+        use super::*;
+        use pretty_assertions::assert_eq;
+
+        /// Empty values returned are unchanged
+        #[test]
+        fn test_empty() {
+            let raw_value = "".as_bytes();
+            let attr = Attribute::from(("foo".as_bytes(), raw_value));
+            assert_eq!(attr.normalized_value(), Ok(Cow::Borrowed("")));
+        }
+
+        /// Already normalized values are returned unchanged
+        #[test]
+        fn test_already_normalized() {
+            let raw_value = "foobar123".as_bytes();
+            let output = "foobar123";
+            let attr = Attribute::from(("foo".as_bytes(), raw_value));
+            assert_eq!(attr.normalized_value(), Ok(Cow::Borrowed(output)));
+        }
+
+        /// Return, tab, and newline characters (0xD, 0x9, 0xA) must be substituted with
+        /// a space character
+        #[test]
+        fn test_space_replacement() {
+            let raw_value = "\r\nfoo\rbar\tbaz\n\ndelta\n".as_bytes();
+            let output = "  foo bar baz  delta ".to_owned();
+            let attr = Attribute::from(("foo".as_bytes(), raw_value));
+            assert_eq!(attr.normalized_value(), Ok(Cow::Owned(output)));
+        }
+
+        /// Entities must be terminated
+        #[test]
+        fn test_unterminated_entity() {
+            let raw_value = "abc&quotdef".as_bytes();
+            let attr = Attribute::from(("foo".as_bytes(), raw_value));
+            assert_eq!(
+                attr.normalized_value(),
+                Err(EscapeError::UnterminatedEntity(3..11))
+            );
+        }
+
+        /// Unknown entities raise error
+        #[test]
+        fn test_unrecognized_entity() {
+            let raw_value = "abc&unkn;def".as_bytes();
+            let attr = Attribute::from(("foo".as_bytes(), raw_value));
+            assert_eq!(
+                attr.normalized_value(),
+                Err(EscapeError::UnrecognizedSymbol(4..8, "unkn".to_owned()))
+                // TODO: UnrecognizedSymbol and UnterminatedEntity have divergent behavior
+                // regarding the start index of the range. Inappropriate?  Existing
+                // unescape code has this same issue.
```

Comment:
@Mingun I can see a potential argument for having the start point of the range be different from each, but I'd just like to check with you, whether you think this is OK or an oversight to be fixed?

Reply:
I also noticed that and at first glance it seems that both errors should cover equivalent ranged. I'd tend to fix that, but I didn't dig deep.